pulsar-metadata/src/main/java/org/apache/pulsar/metadata/bookkeeper/PulsarRegistrationClient.java

```diff
@@ -153,42 +178,57 @@ public void unwatchReadOnlyBookies(RegistrationListener registrationListener) {
+     * local cache in background sequentially.
+     */
+    private void updatedBookies(Notification n) {
+        // make the notification callback run sequential in background.
+        final String path = n.getPath();
+        if (!path.startsWith(bookieReadonlyRegistrationPath) || !path.startsWith(bookieRegistrationPath)) {
+            // ignore unknown path
+            return;
         }
-    }
-
-    private void handleUpdatedBookieNode(Notification n) {
-        BookieId bookieId = stripBookieIdFromPath(n.getPath());
+        final BookieId bookieId = stripBookieIdFromPath(n.getPath());
         if (bookieId != null) {
-            log.info("Bookie {} info updated", bookieId);
-            readBookieServiceInfoAsync(bookieId);
+            log.info("Bookie {} do {}. path: {}", bookieId, n.getType(), n.getPath());
         }
-    }
-
-    private void updatedBookies(Notification n) {
-        if (n.getType() == NotificationType.Created || n.getType() == NotificationType.Deleted) {
-            if (n.getPath().startsWith(bookieReadonlyRegistrationPath)) {
-                getReadOnlyBookies().thenAccept(bookies -> {
-                    readOnlyBookiesWatchers.forEach(w -> executor.execute(() -> w.onBookiesChanged(bookies)));
-                });
-                handleDeletedBookieNode(n);
-            } else if (n.getPath().startsWith(bookieRegistrationPath)) {
-                  getWritableBookies().thenAccept(bookies ->
-                        writableBookiesWatchers.forEach(w -> executor.execute(() -> w.onBookiesChanged(bookies))));
-                handleDeletedBookieNode(n);
-            }
-        } else if (n.getType() == NotificationType.Modified) {
-            if (n.getPath().startsWith(bookieReadonlyRegistrationPath)
-                || n.getPath().startsWith(bookieRegistrationPath)) {
-                handleUpdatedBookieNode(n);
+        sequencer.sequential(() -> {
+            switch (n.getType()) {
+                case Created:
+                    if (path.equals(bookieReadonlyRegistrationPath)) {
+                        return getReadOnlyBookies().thenAccept(bookies ->
+                                readOnlyBookiesWatchers.forEach(w ->
+                                        executor.execute(() -> w.onBookiesChanged(bookies))));
+                    }
+                    return getWritableBookies().thenAccept(bookies ->
+                            writableBookiesWatchers.forEach(w ->
+                                    executor.execute(() -> w.onBookiesChanged(bookies))));
+                case Modified:
+                    if (bookieId == null) {
+                        return completedFuture(null);
+                    }
+                    if (path.equals(bookieReadonlyRegistrationPath)) {
+                        return readBookieInfoAsReadonlyBookie(bookieId).thenApply(__ -> null);
+                    }
+                    return readBookieServiceInfoAsync(bookieId).thenApply(__ -> null);
```

Comment:
Is it also better to remove the cache first? since it has been updated.

Reply:
It is enough to ensure ultimate consistency. If it is deleted first, there is a gap in the cache between the deletion and the new one. :)

Reply:
Ok, got it.